docs/spec_draft.md

```diff
@@ -838,6 +839,74 @@ operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.pad
+
+### Semantics
+
+Expands `operand` by padding around the array as well as between the elements of
+the array with the given `padding_value`.
+
+`edge_padding_low` and `edge_padding_high` specifies the amount of padding added
+at the low-end (next to index 0) and the high-end (next to the highest index) of
+each dimension respectively. The amount of padding can be negative, where the
+absolute value of negative padding indicates the number of elements to remove
+from the specified dimension.
+
+`interior_padding` specifies the amount of padding added between any two
+elements in each dimension which may not be negative. Interior padding occurs
+before edge padding such that negative edge padding will remove elements from
+the interior-padded operand.
+
+### Operands
+
+| Name                | Type                               |
+|---------------------|------------------------------------|
+| `operand`           | tensor of any supported types      |
+| `padding_value`     | tensor of any supported types      |
+| `edge_padding_low`  | 1-dimensional array of type `si64` |
+| `edge_padding_high` | 1-dimensional array of type `si64` |
+| `interior_padding`  | 1-dimensional array of type `si64` |
+
+### Results
+
+| Name     | Type                          |
+|----------|-------------------------------|
+| `result` | tensor of any supported types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the same element type.
+  * (C2) `operand` and `padding_value` have the same element type.
+  * (C3) `padding_value` should be rank 0.
```

Comment:
The table above should mention that the padding value is a scalar.

Reply:
I think this can be inferred from C3, since scalars must be rank 0.

Reply:
Ah I misunderstood you for adding a new constraint. I'll update the table to say:
```
scalar tensor of any supported types
```

Reply:
I don't think that we introduce the notion of a "scalar tensor" in the spec. The closest we have is "Tensors of rank zero, e.g tensor<f32>, have 1 element. Such tensors are allowed and are useful to model scalars", but that's not a canonical definition imo. Let's stick to "0-dimensional tensor"?

A bigger question I think is which constraints should go where. At the moment, it looks like we are specifying element types + ranks of attributes in Operands/Results, and everything else in Constraints. Is the proposal to also specify ranks of tensors in Operands/Result?

Reply:
>Let's stick to "0-dimensional tensor"?

Sure.

>Is the proposal to also specify ranks of tensors in Operands/Result?

Imo, the element types should be in the table since that is what the "Type" part is meant to describe. As for the ranks, similar reasoning. Mostly, the constraints bullet points attempt to match the verification and type inference in ODS and StablehloOps.cpp, at least that's what I've been doing.

I agree that we don't have a clear boundary as to which should go where, but I propose we keep element type and rank stuff in the table description and add them to the Constraints section if the rank or element requirements need to address multiple rows of the Operands table (i.e. `operand` and `result` have the same element type vs. `operand` is a tensor of any supported types). Wdyt?

Reply:
Sounds good!